x-pack/plugins/security/public/management/roles/roles_management_app.tsx

```diff
@@ -0,0 +1,116 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import React from 'react';
+import { render, unmountComponentAtNode } from 'react-dom';
+import { HashRouter as Router, Route, Switch, useParams } from 'react-router-dom';
+import { i18n } from '@kbn/i18n';
+import { CoreSetup, FatalErrorsSetup } from 'src/core/public';
+import { RegisterManagementAppArgs } from '../../../../../../src/plugins/management/public';
+import { SecurityLicense } from '../../../common/licensing';
+import { PluginStartDependencies } from '../../plugin';
+import { UserAPIClient } from '../users';
+import { RolesAPIClient } from './roles_api_client';
+import { RolesGridPage } from './roles_grid';
+import { EditRolePage } from './edit_role';
+import { DocumentationLinksService } from './documentation_links';
+import { IndicesAPIClient } from './indices_api_client';
+import { PrivilegesAPIClient } from './privileges_api_client';
+
+interface CreateParams {
+  fatalErrors: FatalErrorsSetup;
+  license: SecurityLicense;
+  getStartServices: CoreSetup<PluginStartDependencies>['getStartServices'];
+}
+
+export const RolesManagementApp = Object.freeze({
+  id: 'roles',
+  create({ license, fatalErrors, getStartServices }: CreateParams) {
+    return {
+      id: this.id,
+      order: 20,
+      title: i18n.translate('xpack.security.management.rolesTitle', { defaultMessage: 'Roles' }),
+      async mount({ basePath, element, setBreadcrumbs }) {
+        const [
+          { application, docLinks, http, i18n: i18nStart, injectedMetadata, notifications },
+          { data },
+        ] = await getStartServices();
+
+        const rolesBreadcrumbs = [
+          {
+            text: i18n.translate('xpack.security.roles.breadcrumb', { defaultMessage: 'Roles' }),
+            href: `#${basePath}`,
+          },
+        ];
+
+        const rolesAPIClient = new RolesAPIClient(http);
+        const RolesGridPageWithBreadcrumbs = () => {
+          setBreadcrumbs(rolesBreadcrumbs);
+          return <RolesGridPage notifications={notifications} rolesAPIClient={rolesAPIClient} />;
+        };
+
+        const EditRolePageWithBreadcrumbs = ({ action }: { action: 'edit' | 'clone' }) => {
+          const { roleName } = useParams<{ roleName?: string }>();
+
+          setBreadcrumbs([
+            ...rolesBreadcrumbs,
+            action === 'edit' && roleName
+              ? { text: roleName, href: `#${basePath}/edit/${roleName}` }
```

Comment:
In lieu of this suggestion, we can also use the `getEditRoleHref` function within `management_urls` if we update that to respect the `basePath` provided at mount time, which takes care of escaping the role name.
 
```suggestion
              ? { text: roleName, href: `#${basePath}/edit/${encodeURIComponent(roleName)}` }
```

Reply:
Thanks! I think I'll add `encodeURIComponent` explicitly here (and in other management apps) to not depend on `management_urls` with its unclear future :)